File: leetcode/string/l3.java
package leetcode.string;

import java.util.HashMap;
import java.util.Map;

public class l3 {

}

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int ans = 0;
        int start, end;
        Map<Character, Integer> map = new HashMap<>();
        start = end = 0;
        while (end < s.length()) {
            char ch = s.charAt(end);
            if (map.containsKey(ch)) {
                int curIndex = map.get(ch).intValue();
                for (; start < curIndex; start++) {
                    map.remove(s.charAt(start));
                }
            }
            map.put(ch, end);
            ans = Math.max(ans, end - start + 1);
            end += 1;
        }
        return ans;
    }
}

File: leetcode/string/l3.ts

//map
// function _lengthOfLongestSubstring(s: string): number {
//     let ans = 0;
//     let start, end;
//     start = end = 0;
//     let map = new Map<string, number>();
//     while (end < s.length) {
//         let ch = s.charAt(end);
//         if (map.has(ch)) {
//             let curIndex = map.get(ch);
//             for (; start <= curIndex; start++) {
//                 map.delete(ch);
//             }
//         }
//         map.set(ch, end);
//         ans = Math.max(ans, end - start + 1);
//         end += 1;
//     }
//     return ans;
// };

//set
function lengthOfLongestSubstring(s: string): number {
    let ans = 0;
    let start = 0, end = 0;
    let set = new Set<string>();
    while (end < s.length) {
        let ch = s.charAt(end);
        if (set.has(ch)) {
            while (start < end && ch != s.charAt(start)) {
                set.delete(ch);
                start++;
            }
            if (start < end && ch == s.charAt(start)) {
                start++;
            }
        }
        set.add(ch);
        ans = Math.max(ans, end - start + 1);
        end++;
    }
    return ans;
}
